### ISSUE TITLE
How to fix 2 unbound.service warnings: unset environment variable evaluates to an empty string: DAEMON_OPTS and subnetcache: prefetch is set but not working

### ISSUE THREAD

**Issue by sunbearc22** [NLnetLabs/unbound#1148]
**Describe the bug**
unbound.service shows two warnings in yellow colour. How do I resolve them?
1. `unbound.service: Referenced but unset environment variable evaluates to an empty string: DAEMON_OPTS`
2. `warning: subnetcache: prefetch is set but not working for data originating from the subnet module cache.`

![image](https://github.com/user-attachments/assets/2e1ce452-0d07-4ac7-aeb6-d0ef68bfbf8b)

**To reproduce**
1. Install [Ubuntu 24.04.01 Server](https://ubuntu.com/download/server#system-requirements) on an [ACER Aspire One Netbook](https://en.wikipedia.org/wiki/Acer_Aspire_One),
4. Install the Pi-Hole package into Ubuntu 24.01 server via [One-Step Automated Install](https://docs.pi-hole.net/main/basic-install/#one-step-automated-install) . 
5. Install Unbound and set up Pi-hole as a recursive DNS server solution using [Unbound](https://docs.pi-hole.net/guides/dns/unbound/#setting-up-pi-hole-as-a-recursive-dns-server-solution).
6. Run `sudo service unbound status` in terminal.

**Expected behavior**
No warnings from `sudo service unbound status` command.

**System:**
 - Unbound version: 1.19.2
 - OS: Ubuntu Server  24.04.01
 - `unbound -V` output: 
    ```
   Configure line: --build=x86_64-linux-gnu --prefix=/usr --includedir=${prefix}/include --mandir=${prefix}/share/man -- infodir=${prefix}/share/info --sysconfdir=/etc --localstatedir=/var --disable-option-checking --disable-silent-rules --libdir=${prefix}/lib/x86_64-linux-gnu --runstatedir=/run --disable-maintainer-mode --disable-dependency-tracking --with-pythonmodule --with-pyunbound --enable-subnet --enable-dnstap --enable-systemd --enable-cachedb --with-libhiredis --with-libnghttp2 --with-chroot-dir= --with-dnstap-socket-path=/run/dnstap.sock --disable-rpath --with-pidfile=/run/unbound.pid --with-libevent --enable-tfo-client --with-rootkey-file=/usr/share/dns/root.key --enable-tfo-server
   Linked libs: libevent 2.1.12-stable (it uses epoll), OpenSSL 3.0.13 30 Jan 2024
   Linked modules: dns64 python cachedb subnetcache respip validator iterator
   TCP Fastopen feature available
   ```

**Additional information**
```/etc/unbound/unbound.conf.d/pi-hole.conf:```

```
server:
    # If no logfile is specified, syslog is used
    # logfile: "/var/log/unbound/unbound.log"
    verbosity: 0

    interface: 127.0.0.1
    port: 5335
    do-ip4: yes
    do-udp: yes
    do-tcp: yes

    # May be set to yes if you have IPv6 connectivity
    do-ip6: no

    # You want to leave this to no unless you have *native* IPv6. With 6to4 and
    # Terredo tunnels your web browser should favor IPv4 for the same reasons
    prefer-ip6: no

    # Use this only when you downloaded the list of primary root servers!
    # If you use the default dns-root-data package, unbound will find it automatically
    #root-hints: "/var/lib/unbound/root.hints"

    # Trust glue only if it is within the server's authority
    harden-glue: yes

    # Require DNSSEC data for trust-anchored zones, if such data is absent, the zone becomes BOGUS
    harden-dnssec-stripped: yes

    # Don't use Capitalization randomization as it known to cause DNSSEC issues sometimes
    # see https://discourse.pi-hole.net/t/unbound-stubby-or-dnscrypt-proxy/9378 for further details
    use-caps-for-id: no

    # Reduce EDNS reassembly buffer size.
    # IP fragmentation is unreliable on the Internet today, and can cause
    # transmission failures when large DNS messages are sent via UDP. Even
    # when fragmentation does work, it may not be secure; it is theoretically
    # possible to spoof parts of a fragmented DNS message, without easy
    # detection at the receiving end. Recently, there was an excellent study
    # >>> Defragmenting DNS - Determining the optimal maximum UDP response size for DNS <<<
    # by Axel Koolhaas, and Tjeerd Slokker (https://indico.dns-oarc.net/event/36/contributions/776/)
    # in collaboration with NLnet Labs explored DNS using real world data from the
    # the RIPE Atlas probes and the researchers suggested different values for
    # IPv4 and IPv6 and in different scenarios. They advise that servers should
    # be configured to limit DNS messages sent over UDP to a size that will not
    # trigger fragmentation on typical network links. DNS servers can switch
    # from UDP to TCP when a DNS response is too big to fit in this limited
    # buffer size. This value has also been suggested in DNS Flag Day 2020.
    edns-buffer-size: 1232

    # Perform prefetching of close to expired message cache entries
    # This only applies to domains that have been frequently queried
    prefetch: yes

    # One thread should be sufficient, can be increased on beefy machines. In reality for most users running on small networks or on a single machine, it should be unnecessary to seek performance enhancement by increasing num-threads above 1.
    num-threads: 1

    # Ensure kernel buffer is large enough to not lose messages in traffic spikes
    so-rcvbuf: 1m

    # Ensure privacy of local IP ranges
    private-address: 192.168.0.0/16
    private-address: 169.254.0.0/16
    private-address: 172.16.0.0/12
    private-address: 10.0.0.0/8
    private-address: fd00::/8
    private-address: fe80::/10
```

**Comment by gthess** [comment]
```
unbound.service: Referenced but unset environment variable evaluates to an empty string: DAEMON_OPTS
```
We are not packaging Unbound for any system at the moment. This issue is specific to the Unbound package provided by your operating system.
It seems that the unbound.service script installed in your system uses the `DAEMON_OPTS` variable when managing Unbound as a service but there are no actual options defined. This should be fine as the package maintainer probably decided that no extra options need to be defined for the service.

(You could edit the service file and comment out DAEMON_OPTS but I wouldn't recommend messing with system installed files as you will probably face conflicts during updates)

--- 

```
warning: subnetcache: prefetch is set but not working for data originating from the subnet module cache.
```
This is a warning from Unbound itself because both the [`prefetch`](https://unbound.docs.nlnetlabs.nl/en/latest/manpages/unbound.conf.html#unbound-conf-prefetch) option and the subnetcache module are enabled. It is just a user warning to set expectations that cached records from the subnetcache module will not trigger the prefetch logic. The prefetch logic will continue to work for records coming from the global cache.

(Although you have not explicitly activated the subnetcache module, Unbound was compiled with such support and the subnetcache module is implicitly activated. If you don't need subnetcache support you can explicitly set [`module-config: "validator iterator"`](https://unbound.docs.nlnetlabs.nl/en/latest/manpages/unbound.conf.html#unbound-conf-module-config) which is the default setting when subnetcache is not compiled in, or turn off prefetch. Leaving both on is still fine)

---

In short, you don't have to do anything for those warnings :)

**Comment by ddribin** [comment]
FWIW, for those coming here via search (like me), I've found you can avoid the `DAEMON_OPTS` log on Ubuntu 24.04 LTS by adding this to `/etc/default/unbound`:

```
DAEMON_OPTS=""
```

**Comment by sunbearc22** [comment]
I discovered the reason for this msg: `unbound.service: Referenced but unset environment variable evaluates to an empty string: DAEMON_OPTS`

It is caused by `/usr/lib/systemd/system/unbound.service`  (which I think was created by Ubuntu during the installation of `unbound` ):
```
# cat /usr/lib/systemd/system/unbound.service
[Unit]
Description=Unbound DNS server
Documentation=man:unbound(8)
After=network.target
Before=nss-lookup.target
Wants=nss-lookup.target

[Service]
Type=notify
Restart=on-failure
EnvironmentFile=-/etc/default/unbound
ExecStartPre=-/usr/libexec/unbound-helper chroot_setup
ExecStartPre=-/usr/libexec/unbound-helper root_trust_anchor_update
ExecStart=/usr/sbin/unbound -d -p $DAEMON_OPTS
ExecStopPost=-/usr/libexec/unbound-helper chroot_teardown
ExecReload=+/bin/kill -HUP $MAINPID

[Install]
WantedBy=multi-user.target
```

My 2 cents: Essentially, this `ExecStart=/usr/sbin/unbound -d -p $DAEMON_OPTS` command instructs `systemd` to execute `unbound`  with the option `-p`, i.e. do not create a pidfile, but direct the `pid` of the `unbound` process to be store in this variable `DAEMON_OPTS`, which store in `EnvironmentFile=-/etc/default/unbound`. 

This explains why the solution by @ddribin  works. 

**I have a question:** What is the benefit of storing the `pid` of the `unbound` process in an environment variable vs inside a file? Is this for security reason?

**Comment by ddribin** [comment]
While I think you have identified the correct file, it doesn't store the PID in `DAEMON_OPTS`. A command cannot set an environment variable, like that.

It looks to me like this is a way to pass extra arguments to `unbound` when it starts up, if you want. The problem is that the shell expects this variable to be set to *something*. That's why it is complaining about the unset variable. My change sets the variable to an empty string so the `ExecStart` command evaluates to:

```
ExecStart=/usr/sbin/unbound -d -p ""
```

And the empty string gets ignored and interpreted as "no extra arguments".

**Comment by sunbearc22** [comment]
@ddribin  In this case, I wonder why the Ubuntu packaging folks did:
`ExecStart=/usr/sbin/unbound -d -p $DAEMON_OPTS ` 
and not just
`ExecStart=/usr/sbin/unbound -d -p`

**Comment by ddribin** [comment]
Because this allows you to customize the options passed to [`unbound(8)`](https://unbound.docs.nlnetlabs.nl/en/latest/manpages/unbound.html), if you want. They just forgot the critical step of setting it to empty, in case it's not overridden (which is a common case). A bug should probably be opened with Ubuntu to address this.

**Comment by evrial** [comment]
They are simply not intelligent, unbound has no usable commands to pass so this is pure garbage
```
unbound -h
usage:  unbound [options]
        start unbound daemon DNS resolver.
-h      this help.
-c file config file to read instead of /etc/unbound/unbound.conf
        file format is described in unbound.conf(5).
-d      do not fork into the background.
-p      do not create a pidfile.
-v      verbose (more times to increase verbosity).
-V      show version number and build options.
```

**Comment by slappycat2** [comment]
Who let Zuckerberg in here?
